### ISSUE TITLE
ProjectDependenciesCommandFactory fails to create command when assembly name differs from package name

### ISSUE THREAD

**Issue by natemcmaster** [dotnet/cli#2592]
`ProjectDependenciesCommandFactory` cannot create a command to a project dependency if the assembly name is different from the library name.

I believe the source of the issue is that `ProjectDependenciesCommandFactory.Create` a parameter  "commandName".  `ProjectDependenciesCommandResolver` uses to [find the library](https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectDependenciesCommandResolver.cs#L97-105) and `PackagedCommandSpecFactory` uses this name to [find the assembly.](https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactory.cs#L24-25)

This blocks scenarios such as  `dotnet-razor-tooling`  where we need a command to invoke itself inside a project context **and** we want the package name to be different from the assembly name.
 
e.g. Package name: "Microsoft.AspNetCore.Razor.Tooling"
    Assembly name: "dotnet-razor-tooling"

### Sample repro
Run `build.sh` inside this repro project:
[dispatching-tool-repro.zip](https://github.com/dotnet/cli/files/227044/dispatching-tool-repro.zip)

### Error
```
Microsoft.DotNet.Cli.Utils.CommandUnknownException: No executable found matching command "dotnet-my-tool"
```

### Environment
"Microsoft.DotNet.Cli.Utils/1.0.0-rc2-002426"
dotnet 1.0.0-rc2-002439

cc @NTaylorMullen @DamianEdwards @moozzyk

**Comment by moozzyk** [comment]
Look like when the tool calls back into ProjectDependenciesCommandFactory the factory only checks the lock file. Since it is a project reference there are is no target dll name and as a result [this check](https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectDependenciesCommandResolver.cs#L103) fails. I guess for project references the factory needs to load the project.json file to get the actual outputName from compilationOptions.OutputName and compare against it.

**Comment by muratg** [comment]
cc @piotrpMSFT @danroth27

**Comment by moozzyk** [comment]
Actually the repro was not correct - as it was we ended up with a project references and we already know this does not work (https://github.com/dotnet/cli/issues/1215). It still does not work with packages but the fix for this might be simpler.

**Comment by natemcmaster** [comment]
Ah, good catch @moozzyk. Here's the repro changed so the dependency is a package, not a project. The error is still the same.
[dispatching-tool-update1.zip](https://github.com/dotnet/cli/files/228861/dispatching-tool-update1.zip)

**Comment by piotrpMSFT** [comment]
There hasn't been a behavioral change in this space for quite some time. Why is this coming in as blocking so late?

**Comment by moozzyk** [comment]
Because enabling packages whose names were different from the dll came along only last week (https://github.com/dotnet/cli/issues/2322) and this when we started converting the tools.

On the bright side - I think the fix might be a single liner.

**Comment by moozzyk** [comment]
Here is a prototype of the fix: 
https://github.com/dotnet/cli/compare/pawelka/tools#diff-16c83a05b97f578bc7a5c809ad865920R104
I am looking at how to test this.
I am also considering throwing an exception for ambiguous commands - it would be easier to find later when https://github.com/dotnet/cli/issues/2540 is being fixed.

**Comment by brthor** [comment]
@moozzyk I'm not sure this fix is going to work since commandName is also the name of the package which gets restored. Did you have success testing it?

**Comment by moozzyk** [comment]
Yes. I tested this manually and also added a test for that. I did debug the code before and after the change and it fixed the issue.

**Comment by piotrpMSFT** [comment]
@natemcmaster @NTaylorMullen @moozzyk just to be super clear, the goal here is to address only project dependency extensions. `tools` dependencies will continue to maintain their existing behavior. Can you confirm?

**Comment by moozzyk** [comment]
Yes. `tools` dependencies already can handle packages where the dll name is different from the package name. I was able to convert 4 tools already that didn't need project dependencies and they work fine.

**Comment by piotrpMSFT** [comment]
great. Can you add a test that covers that scenario as well so it doesn't get regressed?

**Comment by moozzyk** [comment]
Yes. I will add a test as per https://github.com/dotnet/cli/pull/2620/files#r60618140

**Comment by moozzyk** [comment]
Fixed in 1aac8fb
